### DISCUSSION TITLE
Can anyone else use sodium in PowerShell?

### DISCUSSION THREAD

**Discussion by Szeraax** [jedisct1/libsodium#1277] [Q&A] [answered]
Hi all, I'm trying to use sodium in PowerShell, which is a .NET derivative. Normally with .NET packages in PowerShell, I can use `Add-Type -Path path\to\file.dll` and then gain access to all its .net types, methods, etc. 

I've tried all the download targets that I can. I've tried PowerShell and PowerShell Core (Core is generally preferred). I've used PowerShell x86 and x64 (x64 is generally the preferred one). I've used older versions of sodium (1.17.1). And NONE of them load into powershell for me like I would hope. 

In PowerShell Core, when I run: `Add-Type -PassThru -Path .\libsodium.dll` I get:
`Add-Type: Bad IL format. The format of the file '.\libsodium.dll' is invalid.`

In PowerShell x86 (non-core), when I run `Add-Type -PassThru -Path .\libsodium.dll` (from its folder in Program Files) I get:
`Add-Type : Could not load file or assembly 'file:///C:\Program Files\PackageManagement\NuGet\Packages\libsodium.1.0.18.4\runtimes\win-x86\native\libsodium.dll' or one of its dependencies. The module was expected to contain an assembly manifest.`

I do have the C++ redistributable installed and everything. Just checking to see if anyone sees this same behavior on their systems. I know, its a bit weird to load this into PowerShell, but I just really wasn't expecting this .net assembly to not work here like others do.

**Comment by Szeraax** [comment]
Note: I was directed here instead of the issues bug report area. Hopefully others will be able to confirm if they also are unable to add sodium to a PowerShell session.

**Comment by ektrah** [comment, marked as answer]
`libsodium.dll` is not a .NET assembly. Have a look at the [language bindings for .NET](https://doc.libsodium.org/bindings_for_other_languages).

**Reply by Szeraax** [reply]
Well, I was originally going to open an issue on NSec, since my ultimate goal is to verify an ED25519 signature on a message in PowerShell using NSec's .NET implementation. As far as I can tell, I am able to load the NSec assembly into PowerShell fine, I just can't actually do any verification because the sodium assembly isn't loaded. Should I go file in NSec instead then?

**Reply by Szeraax** [reply]
I figured out that if I have libsodium.dll in the same folder as the NSec.Cryptography.dll, then it will generate keypairs. Sadly, PowerShell Core can't work with spans of bytes (something about no access to stacks), so I will have to try to use another assembly for my task besides NSec. But that's not a bug with NSec or Sodium, so we're all good here. Thank you for the tip, Ektrah!

**Reply by Szeraax** [reply]
As a followup to this task, if anyone ever cares: I was able to use ASodium which has the capability to take byte arrays instead of spans to verify a detached ed25519 signature. [PowerShell code here](https://github.com/Szeraax/trustcircle/blob/main/FunctionApp/profile.ps1#L24-L63)

```ps
    [string]$message = $Request.Headers."x-signature-timestamp" + $Request.RawBody


    [byte[]]$public_bytes = $publicKey -replace "(..)", '0x$1|' -split "\|" | Where-Object { $_ }
    [byte[]]$message_bytes = [System.Text.Encoding]::UTF8.GetBytes($message)
    [byte[]]$signature_bytes = $Request.Headers."x-signature-ed25519" -replace "(..)", '0x$1|' -split "\|" | Where-Object { $_ }


    $result = [ASodium.SodiumPublicKeyAuth]::VerifyDetached(
        $signature_bytes,
        $message_bytes,
        $public_bytes
    )
```
